### PR TITLE
[PaPILO] first attempt at building papilo

### PR DESCRIPTION
**Pull request by matbesancon**
This is a presolver for optimization problems. I got an error running the wizard locally on the latest stable BB:

```
[ 47%] Building CXX object binaries/CMakeFiles/papilo-executable.dir/__/src/papilo.cpp.o
cd /workspace/srcdir/papilo/binaries && /opt/bin/x86_64-linux-gnu-libgfortran3-cxx11/x86_64-linux-gnu-g++ --sysroot=/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/  -DBOOST_ALL_NO_LIB -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_SERIALIZATION_DYN_LINK -DPAPILO_MPS_WRITER -DPAPILO_USE_EXTERN_TEMPLATES -I/workspace/srcdir/papilo/binaries -I/workspace/srcdir/papilo -I/workspace/srcdir/papilo/src -I/workspace/srcdir/papilo/external -I/workspace/srcdir/papilo/external/tbb/include  -Wno-psabi -O3 -DNDEBUG -fPIE -fvisibility=hidden -fvisibility-inlines-hidden   -std=gnu++14 -o CMakeFiles/papilo-executable.dir/__/src/papilo.cpp.o -c /workspace/srcdir/papilo/src/papilo.cpp
[ 49%] Linking CXX executable ../bin/papilo
cd /workspace/srcdir/papilo/binaries && /usr/bin/cmake -E cmake_link_script CMakeFiles/papilo-executable.dir/link.txt --verbose=false
/opt/x86_64-linux-gnu/bin/../lib/gcc/x86_64-linux-gnu/5.2.0/../../../../x86_64-linux-gnu/bin/ld: CMakeFiles/papilo-executable.dir/__/src/papilo.cpp.o: undefined reference to symbol 'clock_gettime@@GLIBC_2.2.5'
/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root//lib64/librt.so.1: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[2]: *** [binaries/CMakeFiles/papilo-executable.dir/build.make:111: bin/papilo] Error 1
make[2]: Leaving directory '/workspace/srcdir/papilo'
make[1]: *** [CMakeFiles/Makefile2:194: binaries/CMakeFiles/papilo-executable.dir/all] Error 2
make[1]: Leaving directory '/workspace/srcdir/papilo'
make: *** [Makefile:161: all] Error 2****
```

### CODE REVIEW

File: P/PaPILO/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+products = [
+    ExecutableProduct("papilo", :papilo),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="boost_jll", uuid="28df3c45-c428-5900-9ff8-a3135698ca75")),
+    Dependency(PackageSpec(name="Bzip2_jll", uuid="6e34b625-4abd-537c-b88f-471c36dfa7a0")),
+    Dependency(PackageSpec(name="CompilerSupportLibraries_jll", uuid="e66e0078-7015-5450-92f7-15fbd957f2ae")),
+    Dependency(PackageSpec(name="GMP_jll", uuid="781609d7-10c4-51f6-84f2-b8444358ff6d"), v"6.1.2"),
```
Comment (giordano) [inline P/PaPILO/build_tarballs.jl:43]:
```suggestion
    Dependency("boost_jll"; compat="=1.76.0"),
    Dependency("Bzip2_jll"; compat="1.0.7"),
    Dependency(PackageSpec(name="CompilerSupportLibraries_jll", uuid="e66e0078-7015-5450-92f7-15fbd957f2ae")),
    Dependency("GMP_jll"; compat="6.1.2"),
```

File: P/PaPILO/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    ARGS,
+    name,
+    version,
+    sources,
+    script,
+    platforms,
+    products,
+    dependencies;
+    preferred_gcc_version = v"5.2.0",
+)
```
Comment (giordano) [inline P/PaPILO/build_tarballs.jl:58]: Can we require Julia v1.6?

Comment (matbesancon) [reply P/PaPILO/build_tarballs.jl:58]: yes no issue with that

File: P/PaPILO/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+build_tarballs(
+    ARGS,
+    name,
+    version,
+    sources,
+    script,
+    platforms,
+    products,
+    dependencies;
+    preferred_gcc_version = v"5.2.0",
```
Comment (giordano) [inline P/PaPILO/build_tarballs.jl:57]: Annoyingly enough, you need at least GCC 7 when linking against our boost build
```suggestion
    preferred_gcc_version = v"7",
```

**Comment by giordano** [comment]
I was trying to debug locally, but the build arrived to eat 25 GiB of memory (15 from my RAM + 10 of swap).  I'm certainly not going to touch this build again myself.

File: P/PaPILO/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/papilo
+mkdir -p build
+cd build
+cmake -DCMAKE_INSTALL_PREFIX=${prefix} \
+    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
+    -DCMAKE_BUILD_TYPE=Release \
+    ..
+make -j${nprocs}
```
Comment (matbesancon) [inline P/PaPILO/build_tarballs.jl:25]:
```suggestion
make
```

**Comment by matbesancon** [comment]
I had the same issue only when running with the parallel make, I removed it for now, still strange though

**Comment by matbesancon** [comment]
ok there are some warnings due to visibility of members and then errors tied to TBB which could be from using a different version (TBB vs oneTBB)

File: P/PaPILO/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    ExecutableProduct("papilo", :papilo),
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency("boost_jll"; compat="=1.76.0"),
+    Dependency("Bzip2_jll"; compat="1.0.7"),
+    Dependency(PackageSpec(name="CompilerSupportLibraries_jll", uuid="e66e0078-7015-5450-92f7-15fbd957f2ae")),
+    Dependency("GMP_jll"; compat="6.1.2"),
+    Dependency(PackageSpec(name="oneTBB_jll", uuid="1317d2d5-d96f-522e-a858-c73665f53c3e")),
```
Comment (matbesancon) [inline P/PaPILO/build_tarballs.jl:44]:
```suggestion
```

**Comment by matbesancon** [comment]
removed oneTBB for now as papilo does not support it yet, only the older TBB 2020 version which was not compiled on Yggdrasil

**Comment by matbesancon** [comment]
the current error is when using the vendored old TBB version:
```
getconf: GNU_LIBC_VERSION: unknown variable
```

**Comment by matbesancon** [comment]
last error that will need fixing is finding quadmath currently fails:
```
[07:17:52] -- Found Boost: /opt/aarch64-linux-gnu/aarch64-linux-gnu/sys-root/usr/local/lib/cmake/Boost-1.76.0/BoostConfig.cmake (found suitable version "1.76.0", minimum required is "1.72")  
[07:17:53] -- Found GMP: /opt/aarch64-linux-gnu/aarch64-linux-gnu/sys-root/usr/local/include  
[07:17:53] -- Performing Test Quadmath_USE_DIRECTLY
[07:17:53] -- Performing Test Quadmath_USE_DIRECTLY - Failed
[07:17:53] -- Could NOT find Quadmath (missing: Quadmath_LIBRARY Quadmath_INCLUDE_DIR) 
```
I couldn't figure out in previous issues whether this could be tied to the GCC version used?

**Comment by matbesancon** [comment]
error when using musl with tbb, it seems to be a different version:
```
[11:21:04] [ 51%] Linking CXX executable ../bin/papilo
[11:21:04] cd /workspace/srcdir/papilo/build/binaries && /usr/bin/cmake -E cmake_link_script CMakeFiles/papilo-executable.dir/link.txt --verbose=true
[11:21:04] /opt/bin/aarch64-linux-musl-libgfortran4-cxx11/aarch64-linux-musl-g++ --sysroot=/opt/aarch64-linux-musl/aarch64-linux-musl/sys-root/   -Wno-psabi -O3 -DNDEBUG   CMakeFiles/papilo-executable.dir/__/src/papilo.cpp.o  -o ../bin/papilo  -Wl,-rpath,/workspace/destdir/lib:/usr/local/lib ../libpapilo-core.a /workspace/destdir/lib/libboost_iostreams.so.1.76.0 /workspace/destdir/lib/libboost_serialization.so.1.76.0 /workspace/destdir/lib/libboost_program_options.so.1.76.0 /opt/aarch64-linux-musl/aarch64-linux-musl/sys-root/usr/local/lib/libtbb.so.12.2 -lrt /opt/aarch64-linux-musl/aarch64-linux-musl/sys-root/usr/local/lib/libgmp.so /opt/aarch64-linux-musl/aarch64-linux-musl/sys-root/usr/local/lib/libgmpxx.so 
[11:21:05] /opt/aarch64-linux-musl/aarch64-linux-musl/sys-root/usr/local/lib/libtbb.so.12.2: undefined reference to `swapcontext'
[11:21:05] /opt/aarch64-linux-musl/aarch64-linux-musl/sys-root/usr/local/lib/libtbb.so.12.2: undefined reference to `getcontext'
[11:21:05] /opt/aarch64-linux-musl/aarch64-linux-musl/sys-root/usr/local/lib/libtbb.so.12.2: undefined reference to `makecontext'
[11:21:05] collect2: error: ld returned 1 exit status
[11:21:05] make[2]: *** [binaries/CMakeFiles/papilo-executable.dir/build.make:111: bin/papilo] Error 1
[11:21:05] make[2]: Leaving directory '/workspace/srcdir/papilo/build'
[11:21:05] make[1]: *** [CMakeFiles/Makefile2:194: binaries/CMakeFiles/papilo-executable.dir/all] Error 2
[11:21:05] make[1]: Leaving directory '/workspace/srcdir/papilo/build'
[11:21:05] make: *** [Makefile:161: all] Error 2
```

**Comment by matbesancon** [comment]
Update: this is now building both SCIP with PaPILO and PaPILO itself as an executable. We are still restricted in the platforms by the ones TBB builds on.
Since GCG also fails on windows for the moment, it will be built here along with SCIP and PaPILO

File: S/SCIP_PaPILO/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+make install
+
+mkdir -p ../papilo/build
+cd ../papilo/build
+cmake -DCMAKE_INSTALL_PREFIX=${prefix} \
+    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TARGET_TOOLCHAIN} \
+    -DCMAKE_BUILD_TYPE=Release \
+    ..
+make
+cp bin/papilo ${prefix}/bin/
```
Comment (giordano) [inline S/SCIP_PaPILO/build_tarballs.jl:36]:
```suggestion
cp bin/papilo "${bindir}/papilo${exeext}"
```

File: S/SCIP_PaPILO/build_tarballs.jl [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+mkdir -p ${prefix}/share/licenses/SCIP_PaPILO
+for dir in papilo scip soplex; do
+    cp $WORKSPACE/srcdir/scipoptsuite*/${dir}/COPYING ${prefix}/share/licenses/SCIP_PaPILO/LICENSE_${dir}
+done
+cp $WORKSPACE/srcdir/scipoptsuite*/gcg/LICENSE ${prefix}/share/licenses/SCIP_PaPILO/LICENSE_gcg
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = expand_gfortran_versions(expand_cxxstring_abis(supported_platforms()))
```
Comment (giordano) [inline S/SCIP_PaPILO/build_tarballs.jl:47]: Did you try building for aarch64-apple-darwin?
```suggestion
platforms = expand_gfortran_versions(expand_cxxstring_abis(supported_platforms(; experimental=true)))
```

**Comment by matbesancon** [comment]
It still needs GMP and Bzlib2 to be built with experimental

**Comment by matbesancon** [comment]
Edit: GMP does? https://github.com/JuliaBinaryWrappers/GMP_jll.jl/releases/tag/GMP-v6.2.1%2B1
where does the warning come from

File: S/SCIP_PaPILO/build_tarballs.jl [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    LibraryProduct("libscip", :libscip),
+    LibraryProduct("libgcg", :libgcg),
+]
+
+dependencies = [
+    Dependency(PackageSpec(name="bliss_jll", uuid="508c9074-7a14-5c94-9582-3d4bc1871065")),
+    Dependency(PackageSpec(name="boost_jll", uuid="28df3c45-c428-5900-9ff8-a3135698ca75"); compat="=1.76.0"),
+    Dependency(PackageSpec(name="Bzip2_jll", uuid="6e34b625-4abd-537c-b88f-471c36dfa7a0"); compat="1.0.7"),
+    Dependency(PackageSpec(name="CompilerSupportLibraries_jll", uuid="e66e0078-7015-5450-92f7-15fbd957f2ae")),
+    Dependency(PackageSpec(name="GMP_jll", uuid="781609d7-10c4-51f6-84f2-b8444358ff6d"), v"6.1.2"),
```
Comment (giordano) [inline S/SCIP_PaPILO/build_tarballs.jl:66]:
```suggestion
    Dependency(PackageSpec(name="GMP_jll", uuid="781609d7-10c4-51f6-84f2-b8444358ff6d"), v"6.2.0"),
```

File: S/SCIP_PaPILO/build_tarballs.jl [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    ExecutableProduct("papilo", :papilo),
+    ExecutableProduct("scip", :scip),
+    LibraryProduct("libscip", :libscip),
+    LibraryProduct("libgcg", :libgcg),
+]
+
+dependencies = [
+    Dependency(PackageSpec(name="bliss_jll", uuid="508c9074-7a14-5c94-9582-3d4bc1871065")),
+    Dependency(PackageSpec(name="boost_jll", uuid="28df3c45-c428-5900-9ff8-a3135698ca75"); compat="=1.76.0"),
+    Dependency(PackageSpec(name="Bzip2_jll", uuid="6e34b625-4abd-537c-b88f-471c36dfa7a0"); compat="1.0.7"),
```
Comment (giordano) [inline S/SCIP_PaPILO/build_tarballs.jl:64]:
```suggestion
    Dependency(PackageSpec(name="Bzip2_jll", uuid="6e34b625-4abd-537c-b88f-471c36dfa7a0"); compat="1.0.8"),
```

**Comment by matbesancon** [comment]
The error seems to be in the "catch" dependency?

https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=15041&view=logs&j=97d3c30b-cdc8-5503-c0f0-063c9cef4178&t=1d68f470-a03d-5572-ece9-bbb324f9e913&l=7459

```
[10:48:58] In file included from /workspace/srcdir/scipoptsuite-8.0.0/papilo/test/TestMain.cpp:26:
[10:48:58] /workspace/srcdir/scipoptsuite-8.0.0/papilo/external/catch/catch.hpp:4939:13: error: unrecognized instruction mnemonic, did you mean: bit, cnt, hint, ins, not?
[10:48:58]             CATCH_BREAK_INTO_DEBUGGER();
[10:48:58]             ^
[10:48:58] /workspace/srcdir/scipoptsuite-8.0.0/papilo/external/catch/catch.hpp:4661:75: note: expanded from macro 'CATCH_BREAK_INTO_DEBUGGER'
[10:48:58]     #define CATCH_BREAK_INTO_DEBUGGER() if( Catch::isDebuggerActive() ) { CATCH_TRAP(); }
[10:48:58]                                                                           ^
[10:48:58] /workspace/srcdir/scipoptsuite-8.0.0/papilo/external/catch/catch.hpp:4640:34: note: expanded from macro 'CATCH_TRAP'
[10:48:58]     #define CATCH_TRAP() __asm__("int $3\n" : : ) /* NOLINT */
[10:48:58]                                  ^
[10:48:58] <inline asm>:1:2: note: instantiated into assembly here
[10:48:58]         int $3
[10:48:58]         ^
[10:48:58] 1 error generated.
[10:48:58] make[2]: *** [papilo/test/CMakeFiles/unit_test.dir/build.make:83: papilo/test/CMakeFiles/unit_test.dir/TestMain.cpp.o] Error 1
[10:48:58] make[2]: *** Waiting for unfinished jobs....
```

**Comment by giordano** [comment]
These are tests, can we disable them?

**Comment by matbesancon** [comment]
good question, I'm looking now but my makefu is failing me

**Comment by matbesancon** [comment]
I'll ping the developers tomorrow

**Pull request merged**